src/render/websocket.js

```diff
@@ -0,0 +1,124 @@
+const headers = require('./headers')
+const cookies = require('./cookies')
+const address = require('./address')
+const object = require('./object')
+const { websocketFactor: makeWebsocketFactor } = require('../make')
+
+function websocket(spec) {
+  let factor = getFactor(spec.request, spec)
+  return render(factor)
+}
+
+function getFactor(request, spec) {
+  const factor = makeWebsocketFactor()
+  if (spec.addSleep) {
+    factor.timeAlive = timeAlive(spec.timeConnected, spec.webSocketMessages)
+    factor.addSleep = spec.addSleep
+    factor.messages = changeMessageTimes(spec.webSocketMessages)
+  }
+  else{
+    factor.messages = spec.webSocketMessages
+  }
+  factor.call = 'connect'
+  address(request, factor)
+  factor.headers = headers(request.headers)
+  factor.cookies = cookies(request.cookies)
+  factor.options = options(factor)
+  factor.args = args(factor)
+
+  return factor
+}
+
+function changeMessageTimes(messages) {
+  /// calculates when to sleep between messages
+  let firstTimeStartsAtZero = 0
+  let firstIndexSetToZero = 0
+  let messagesWithNewTimes = JSON.parse(JSON.stringify(messages))
+  messagesWithNewTimes[firstIndexSetToZero].time = firstTimeStartsAtZero
+  for (let i = 1; i < messages.length; i++){
+    messagesWithNewTimes[i].time = messages[i].time - messages[i - 1].time
+  }
+  return messagesWithNewTimes
+}
+
+function timeAlive(timeConnected, messages) {
+  /// calculates length of time when websocket is open minus the total time the messages took to run before closeing the socket
+  let firstMessageTime = messages[0].time
+  let finalMessageTime = messages[messages.length - 1].time
```

Comment:
nit: Consider checking that messages has at least 1 item

Reply:
Good IDEA i also will change function changeMessageTimes on line 32